client/README.md

```diff
@@ -0,0 +1,1013 @@
+
+* Ensure that `proof.leaf` is empty or 32-bytes.
+* Split `proof.leaf` in two 32-byte arrays:
+    - the first part is a key
+    - the second part is a value_hash
+    - `leaf = None | Some((key, value_hash))`
+
+* Ensure that the byte-length of `proof.bitmap` is smaller or equal to 32 bytes.
+* Ensure that the last byte of `proof.bitmap` is not zero
+* Ensure that the number of bit set to 1 in `proof.bitmap` is equal to the length of `proof.non_default_siblings`.
+
+* Ensure that each item of proof.non_default_siblings` is of size 32-byte.
+
+* If `element_index` is set
+    - if `proof.leaf` is `Some((key, value_hash)`
+        - ensure that `element_key == proof_key`
+        - ensure that `hash == proof_value_hash`
+    - if `proof.leaf` is `None`, abort.
+* If `element_index` is not set
+    - if `proof.leaf` is `Some((key, value_hash)`
+        - ensure that `element_key != proof_key`
+        - ensure that `element_key.common_prefix_bits_len(proof_key) >= siblings.len()`
+    - if `proof.leaf` is `None`, continue.
+
+Proof Verification:
+
+* If `proof.leaf` is `None`, set `current_hash` to `SPARSE_MERKLE_PLACEHOLDER_HASH`.
+* If `proof.leaf` is `Some((key, value_hash)` set `current_hash` to the output of `SHA-3-256(SparseMerkleLeafHasher, HASH_SUFFIX, key, value_hash)`.
+
+```py
+bitmap = Bitmap32(bitmap=proof.bitmap, path=element_key)
+loop {    
+    default, left = bitmap.Next()
+    if default == end {
+        break
+    }
+    if default {
+        sibling = SPARSE_MERKLE_PLACEHOLDER_HASH
+    } else {
+        sibling = proof.non_default_siblings.pop()
+    }
+
+    if left {
+        current_hash = SHA-3-256(SparseMerkleInternal, HASH_SUFFIX, sibling, current_hash)
+    } else {
+        current_hash = SHA-3-256(SparseMerkleInternal, HASH_SUFFIX, current_hash, sibling)
+    }
+}
+
+if current_hash == expected_root_hash {
+    return true
+} else {
+    return false
+}
+```
+
+## SubmitTransaction Request
+
+A client can submit a transaction to the network by encapsulating its LCS (TODO: definition of LCS) encoding into a `SignedTransaction` message wrapped into a `SubmitTransactionRequest`.
+
+```proto
+message SubmitTransactionRequest {
+  types.SignedTransaction signed_txn = 1;
+}
+
+message SignedTransaction {
+    bytes signed_txn = 5;
+}
+```
+
+## SubmitTransaction Response
+
+A response to a `SubmitTransaction` request will either contain some error codes, indicating that the transaction was not successfully processed by the validator, or nothing indicating success.
+
+**This by no mean implies that the transaction was included on the chain**. 
+At this point, a client must poll the network in order to verify that its transaction was successfuly processed.
+TODO: how? GetAccountTransactionBySequenceNumberRequest? Or better with events?
+
+This section specifies how to validate a response to a `SubmitTransactionRequest` message.
+
+The response is a `SubmitTransactionResponse` message as specified below:
+
+```proto
```

Comment:
My biggest concern with a document like this is it has a high likelihood of falling out of date very quickly. It would be much better if instead we maybe linked to the definition?

Reply:
linking to the definition also has its drawbacks: the document is not self-contained anymore, you have to do back-and-forth between the spec and the code to understand it, and you can also end up with a dangling link.